### PR TITLE
fix(DSCI): change serviceMesh and trustCAbundle to pointer type

### PR DESCRIPTION
**Pull request by zdtsw**
- change trustCABundle to pointer, so if not set it is the same as set to Removed
- change serviceMesh to pointer too,  so if not set it is the same as set to Removed
- also include a fix in e2e test which set to use the wrong component name
ref : https://issues.redhat.com/browse/RHOAIENG-6295 

When user install operator from OperatorHub UI it is not a problem, the default fields are added there by CSV
When user manually create DSCI via CLI, they might not know if any new fields have been added into release.
e.g  a mini config of DSCI will cause operator crash
```
apiVersion: dscinitialization.opendatahub.io/v1
kind: DSCInitialization
metadata:
  name: default-dsci
spec:
  applicationsNamespace: opendatahub
  monitoring:
    managementState: Managed
    namespace: opendatahub
```

## How Has This Been Tested?

NEW:
live-build quay.io/wenzhou/opendatahub-operator-catalog:v2.10.0-6
- install opreator
- manually create DSCI CR with out `trustedCABundle`:
```
spec:
  applicationsNamespace: opendatahub
  monitoring:
    managementState: Managed
    namespace: opendatahub
  serviceMesh:
    controlPlane:
      metricsCollection: Istio
      name: data-science-smcp
      namespace: istio-system
    managementState: Managed
```
- no odh-trusted-ca-bundle created into new namespace
- old namespace which had odh-trusted-ca-bundle (created by previous runs) got removed
- no error in operator pod

OLD:
live-build: 'quay.io/wenzhou/opendatahub-operator-catalog:v2.9.0-885'
## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: pkg/trustedcabundle/trustedcabundle.go [hunk 1]
```diff
@@ -163,33 +163,42 @@ func IsTrustedCABundleUpdated(ctx context.Context, cli client.Client, dscInit *d
...
+				return err
+			}
 
-		if istrustedCABundleUpdated || managementStateChanged {
-			err = AddCABundleConfigMapInAllNamespaces(ctx, cli, dscInit)
+			if istrustedCABundleUpdated || managementStateChanged {
+				err = AddCABundleConfigMapInAllNamespaces(ctx, cli, dscInit)
+				if err != nil {
+					log.Error(err, "error adding configmap to all namespaces", "name", CAConfigMapName)
+					return err
```
Comment (bartoszmajsak) [inline pkg/trustedcabundle/trustedcabundle.go:179]: I think we should stick to the principle of handle or throw. In this case we do both - log (handle) and return (throw).

```suggestion
					fmt.Errorf("failed adding configmap %s to all namespaces: %w", CAConfigMapName, err)
```

File: pkg/trustedcabundle/trustedcabundle.go [hunk 2]
```diff
@@ -163,33 +163,42 @@ func IsTrustedCABundleUpdated(ctx context.Context, cli client.Client, dscInit *d
...
+				}
+			}
+		case operatorv1.Removed:
+			log.Info("Trusted CA Bundle injection is set to `Removed` state. Reconciling to delete all cert configmaps")
+			err := RemoveCABundleConfigMapInAllNamespaces(ctx, cli)
 			if err != nil {
-				log.Error(err, "error adding configmap to all namespaces", "name", CAConfigMapName)
+				log.Error(err, "error deleting configmap from all namespaces", "name", CAConfigMapName)
 				return err
```
Comment (bartoszmajsak) [inline pkg/trustedcabundle/trustedcabundle.go:187]: Same principle as above

```suggestion
				fmt.Errorf("error deleting configmap %s from all namespaces", CAConfigMapName, err)
```

File: pkg/trustedcabundle/trustedcabundle.go [hunk 3]
```diff
@@ -163,33 +163,42 @@ func IsTrustedCABundleUpdated(ctx context.Context, cli client.Client, dscInit *d
...
 
-	case operatorv1.Unmanaged:
-		log.Info("Trusted CA Bundle injection is set to `Unmanaged` state. Cert configmaps are no longer managed by DSCI")
+		return nil
+	}
+	log.Info("Trusted CA Bundle is not configed in DSCI, same as default to `Removed` state. Reconciling to delete all cert configmaps")
+	err := RemoveCABundleConfigMapInAllNamespaces(ctx, cli)
+	if err != nil {
+		log.Error(err, "error deleting configmap from all namespaces", "name", CAConfigMapName)
+		return err
```
Comment (bartoszmajsak) [inline pkg/trustedcabundle/trustedcabundle.go:200]:
```suggestion
		fmt.Errorferr, "error deleting configmap %s from all namespaces: %w", CAConfigMapName, err)
```

File: controllers/dscinitialization/servicemesh_setup.go [hunk 4]
```diff
@@ -7,30 +7,37 @@ import (
...
+	emptySpec := infrav1.ServiceMeshSpec{}
+	if instance.Spec.ServiceMesh != emptySpec {
+		switch instance.Spec.ServiceMesh.ManagementState {
+		case operatorv1.Managed:
+			serviceMeshFeatures := feature.ClusterFeaturesHandler(instance, configureServiceMeshFeatures())
+
+			if err := serviceMeshFeatures.Apply(); err != nil {
+				r.Log.Error(err, "failed applying service mesh resources")
+				r.Recorder.Eventf(instance, corev1.EventTypeWarning, "DSCInitializationReconcileError", "failed applying service mesh resources")
+				return err
```
Comment (bartoszmajsak) [inline controllers/dscinitialization/servicemesh_setup.go:25]:
```suggestion
				r.Recorder.Eventf(instance, corev1.EventTypeWarning, "DSCInitializationReconcileError", "failed applying service mesh resources")
				return fmt.Errorf(""failed applying service mesh resources: %w", err)
```

but also you need to import `"fmt"`

File: controllers/dscinitialization/servicemesh_setup.go [hunk 5]
```diff
@@ -7,30 +7,37 @@ import (
...
+			}
 		}
-	}
 
-	return nil
+		return nil
+	}
+	r.Log.Info("ServiceMesh is not configed in DSCI, same as default to 'Removed', existing ServiceMesh CR (owned by operator) will be removed")
+	err := r.removeServiceMesh(instance)
+	return err
```
Comment (bartoszmajsak) [inline controllers/dscinitialization/servicemesh_setup.go:40]:
```suggestion
	return r.removeServiceMesh(instance)
```

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]

**Comment by zdtsw** [comment]
missed your comments, let me fix these tomorrow 🤔

File: bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml [hunk 6]
```diff
@@ -183,7 +183,8 @@ spec:
                           This can be set to 'Serverless' or 'RawDeployment'. The
                           value specified in this field will be used to set the default
                           deployment mode in the 'inferenceservice-config' configmap
-                          for Kserve
+                          for Kserve If no default deployment mode is specified, Kserve
+                          will use Serverless mode
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml]: this is done, i hope

File: controllers/dscinitialization/servicemesh_setup.go [hunk 7]
```diff
@@ -7,30 +7,37 @@ import (
...
-			return err
-		}
-	case operatorv1.Unmanaged:
-		r.Log.Info("ServiceMesh CR is not configured by the operator, we won't do anything")
-	case operatorv1.Removed:
-		r.Log.Info("existing ServiceMesh CR (owned by operator) will be removed")
-		if err := r.removeServiceMesh(instance); err != nil {
-			return err
+	emptySpec := infrav1.ServiceMeshSpec{}
+	if instance.Spec.ServiceMesh != emptySpec {
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply controllers/dscinitialization/servicemesh_setup.go]: either do i,
and from my description in original PR, i said "keep servicemesh but only add a check" not sure what was the decision behind this.
but now i changed it (probably i dont wanna introduce too much and break something :rage4: )

File: pkg/trustedcabundle/trustedcabundle.go (hunk 1 above)
Comment (zdtsw) [reply pkg/trustedcabundle/trustedcabundle.go]: these all should have been changed to return fmt.Errorf(...)

**Comment by zdtsw** [comment]
current PR includes the fix from https://github.com/opendatahub-io/opendatahub-operator/pull/920
if https://github.com/opendatahub-io/opendatahub-operator/pull/920 get merged first we can revert last commits in the PR

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]

File: components/kserve/kserve.go [hunk 8]
```diff
@@ -50,8 +50,8 @@ type Kserve struct {
 	// Mesh (Istio) is prerequisite, since it is used as networking layer.
 	Serving infrav1.ServingSpec `json:"serving,omitempty"`
 	// Configures the default deployment mode for Kserve. This can be set to 'Serverless' or 'RawDeployment'.
-	// The value specified in this field will be used to set the default deployment mode in the 'inferenceservice-config' configmap for Kserve
-	// If no default deployment mode is specified, Kserve will use Serverless mode
+	// The value specified in this field will be used to set the default deployment mode in the 'inferenceservice-config' configmap for Kserve.
+	// This field is optional. If no default deployment mode is specified, Kserve will use Serverless mode.
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply components/kserve/kserve.go:54]: i will prefer to leave this one to serving team.
to update the comments is only to fix the strange API description. but i can remove this "This field is optional" and leave it for another PR to deal with `DefaultDeploymentMode`

File: controllers/dscinitialization/servicemesh_setup.go [hunk 9]
```diff
@@ -12,37 +13,42 @@ import (
 )
 
 func (r *DSCInitializationReconciler) configureServiceMesh(instance *dsciv1.DSCInitialization) error {
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply controllers/dscinitialization/servicemesh_setup.go:15]: i think in the case of `nil` VS `Removed`, the only difference is the message sends to user.
 they may not be aware of  what `nil`/'not set' really means here
Not every user will go and read API to understand the usage and always ask for clarification in "documentation" or some logs info. 

another part is:
i think nil == `Removed`, is this correct?
from "suggestion" code, it feels like nil == `Unmanaged`?  from `serviceMeshManagementState := operatorv1.Unmanaged`

[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply controllers/dscinitialization/servicemesh_setup.go:15]: > i think in the case of nil VS Removed, the only difference is the message sends to user.

I would assume `nil` == not existing so `Removed`. Also I would expect removing this part from the .spec in YAML is the same as marking as `Removed`. So I am not sure it would be surprising for the user. Sounds quite natural **to me** (but I might be biased).

> i think nil == Removed, is this correct?

YES! I was half awake. Corrected. I meant `Removed` of course :)

[the remark this answers is not in the document]
Comment (zdtsw) [reply controllers/dscinitialization/servicemesh_setup.go:15]: I did some updates based on the correct suggestion :wink: 
but also prints "nil" case in the message.

**Review by bartoszmajsak — DISMISSED** [review]
From the "polish" side of things LGTM.

I added @israel-hdez to review the reconcile flow as that is not what I'm very familiar with.

File: components/kserve/kserve.go [hunk 10]
```diff
@@ -103,34 +103,33 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, owne
...
 	if !enabled {
-		if err := k.removeServerlessFeatures(dscispec); err != nil {
-			return err
-		}
-	} else {
-		// Configure dependencies
-		if err := k.configureServerless(cli, dscispec); err != nil {
+		err := k.removeServerlessFeatures(dscispec)
+		// should return here, why continue?
+		return err
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply components/kserve/kserve.go:116]: i added this comments , just to clarify 
in the old code:
if we disable kserve, we run `removeServerlessFeatures` function and only when it is err, we return 
then continue with `deploy.DeployManifestsFromPath` etc
so my question is:
if we disable kserve, we run `removeServerlessFeatures` function , should it return as  well when there is no err, right?

[the remark this answers is not in the document]
Comment (zdtsw) [reply components/kserve/kserve.go:116]: after think twice,  i understand why we did this in the past.
we do need to continue with deploy.DeployManifestsFromPath() and so on, that is in order to pass the "false" to remove deployed manifests from kserve.  
i will revert this part

[the remark this answers is not in the document]
Comment (zdtsw) [reply components/kserve/kserve.go]: ping @israel-hdez here for another round of eyes

File: components/kserve/kserve.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (israel-hdez) [reply components/kserve/kserve.go:54]: The `DefaultDeploymentMode` is just an alias to `string`. Given the `omitempty`, I think it is truly optional unless it is not possible to assign the empty string.

See a screenshot of the resource here: https://github.com/opendatahub-io/opendatahub-operator/pull/864#discussion_r1503288913

I don't recall seeing the field on my own testings... and I do remember trying it myself. So, despite Vedant's screenshot does not give enough additional lines for context, that screenshot should be a complete definition of the kserve component. Notice the `defaultDeploymentMode` field is missing.

File: components/kserve/kserve.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (israel-hdez) [reply components/kserve/kserve.go]: > we do need to continue with deploy.DeployManifestsFromPath() and so on

Yes, this was it, IIRC. Otherwise, my understanding is that some clean-up would not be done.

**Review by israel-hdez — COMMENTED** [review]
I'm not able to see any unwanted impact for model serving.

File: components/kserve/kserve.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (zdtsw) [reply components/kserve/kserve.go]: thanks for confirming

[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply components/kserve/kserve.go]: > we do need to continue with deploy.DeployManifestsFromPath() and so on, that is in order to pass the "false" to remove deployed manifests from kserve.

this is why I added a note in my previous comment that it might be good to simplify the flow here :)

File: components/kserve/kserve.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply components/kserve/kserve.go:54]: The screenshot is truncated so I am not sure that the whole picture, but I assumed it's  a struct though now sure why as the name of the type suggests it might just be alias to the string. Sorry for the noise :)

File: components/kserve/kserve.go [hunk 11]
```diff
@@ -103,7 +103,7 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, owne
 		"odh-model-controller": "RELATED_IMAGE_ODH_MODEL_CONTROLLER_IMAGE",
 	}
 
-	enabled := k.GetManagementState() == operatorv1.Managed
+	enabled := k.GetManagementState() == operatorv1.Managed && dscispec.ServiceMesh != nil
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply components/kserve/kserve.go:106]: I agree, we should keep the original evaluation here - the state of the component as such. We are anyway handling service mesh setup separately where we check this.

File: controllers/dscinitialization/servicemesh_setup.go [hunk 12]
```diff
@@ -12,37 +13,42 @@ import (
...
-			r.Recorder.Eventf(instance, corev1.EventTypeWarning, "DSCInitializationReconcileError", "failed applying service mesh resources")
-			return err
-		}
-	case operatorv1.Unmanaged:
-		r.Log.Info("ServiceMesh CR is not configured by the operator, we won't do anything")
-	case operatorv1.Removed:
-		r.Log.Info("existing ServiceMesh CR (owned by operator) will be removed")
-		if err := r.removeServiceMesh(instance); err != nil {
-			return err
+	if instance.Spec.ServiceMesh != nil {
```
Comment (bartoszmajsak) [inline controllers/dscinitialization/servicemesh_setup.go:16]: So we are keeping the nested `if`? I am confused now as there's also this comment https://github.com/opendatahub-io/opendatahub-operator/pull/885/files#r1535380592 :)

File: pkg/trustedcabundle/trustedcabundle.go [hunk 13]
```diff
@@ -134,31 +134,36 @@ func IsTrustedCABundleUpdated(ctx context.Context, cli client.Client, dscInit *d
...
+			log.Info("Trusted CA Bundle injection is set to `Unmanaged` state. " + CAConfigMapName + " configmaps are no longer managed by operator")
 		}
 
-	case operatorv1.Unmanaged:
-		log.Info("Trusted CA Bundle injection is set to `Unmanaged` state. " + CAConfigMapName + " configmaps are no longer managed by operator")
+		return nil
 	}
 
+	log.Info("Trusted CA Bundle is not configed in DSCI, same as default to `Removed` state. Reconciling to delete all " + CAConfigMapName)
+	if err := RemoveCABundleConfigMapInAllNamespaces(ctx, cli).ErrorOrNil(); err != nil {
```
Comment (bartoszmajsak) [inline pkg/trustedcabundle/trustedcabundle.go:164]: Not part of this PR, but shouldn't `RemoveCABundleConfigMapInAllNamespaces` return an error instead?

File: components/kserve/kserve.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (zdtsw) [reply components/kserve/kserve.go]: Agreed, 
this part need some refactor, even me forgot the original usage of this "enable" flag :rage4:  
shame on my memory

[the remark this answers is not in the document]
Comment (zdtsw) [reply components/kserve/kserve.go]: create https://issues.redhat.com/browse/RHOAIENG-4865 to follow up

File: components/kserve/kserve.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (zdtsw) [reply components/kserve/kserve.go:106]: thanks! reverted this part.

File: controllers/dscinitialization/servicemesh_setup.go (hunk 12 above)
Comment (zdtsw) [reply controllers/dscinitialization/servicemesh_setup.go:16]: looks like i did not check the changes after our conversation :/

but
in order to print this` "ServiceMesh is not configed in DSCI, same as default to 'Removed',"  `will need a `nil` check, or ? :thinking:

File: pkg/trustedcabundle/trustedcabundle.go (hunk 13 above)
Comment (zdtsw) [reply pkg/trustedcabundle/trustedcabundle.go]: updated

**Comment by VaishnaviHire** [comment]
I am not sure if we should move trustedCABundle to pointer. What will be the customCABundle value?

**Comment by zdtsw** [comment]
> I am not sure if we should move trustedCABundle to pointer. What will be the customCABundle value?

customCABundle will keep being as string.
can be "" or user can set their CA bundle value to it.
if use do not set `dscInit.Spec.TrustedCABundle`, it is the same as in the current code with `dscInit.Spec.TrustedCABundle.ManagementState` to `Removed`. 

this is mainly from one of the user for serving team when they do not need anything except using rawdeployment,
for now they need to config TrustedCABundle in default-dsci and set to Removed
if it is doing as pointer, then they can keep their old default-dsci without changing it.

**Comment by zdtsw** [comment]
i think it is kinda related to the comments made in https://issues.redhat.com/browse/RHOAIENG-6261 
"I just upgraded the cluster, so I didn't touch the DSCI CR. When looking at the logs, it complains about missing `spec.trustedCABundle.customCABundle`, but the entire `spec.trustedCABundle` block was missing upon the upgrade"

File: controllers/dscinitialization/servicemesh_setup.go [hunk 14]
```diff
@@ -46,31 +53,32 @@ func (r *DSCInitializationReconciler) configureServiceMesh(instance *dsciv1.DSCI
...
 	return nil
 }
 
 func (r *DSCInitializationReconciler) removeServiceMesh(instance *dsciv1.DSCInitialization) error {
 	// on condition of Managed, do not handle Removed when set to Removed it trigger DSCI reconcile to clean up
-	if instance.Spec.ServiceMesh.ManagementState == operatorv1.Managed {
-		capabilities := []*feature.HandlerWithReporter[*dsciv1.DSCInitialization]{
-			r.serviceMeshCapability(instance, serviceMeshCondition(status.RemovedReason, "Service Mesh removed")),
-		}
+	if instance.Spec.ServiceMesh != nil {
```
Comment (bartoszmajsak) [inline controllers/dscinitialization/servicemesh_setup.go:61]: Can't we make a guard ifs instead of nesting across the whole change? That might be a bit easier to follow.

```suggestion
	if instance.Spec.ServiceMesh == nil {
		return nil
	}	
```

File: controllers/dscinitialization/servicemesh_setup.go [hunk 15]
```diff
@@ -16,7 +16,14 @@ import (
 )
 
 func (r *DSCInitializationReconciler) configureServiceMesh(instance *dsciv1.DSCInitialization) error {
-	switch instance.Spec.ServiceMesh.ManagementState {
+	serviceMeshManagementState := operatorv1.Removed
+	if instance.Spec.ServiceMesh != nil {
```
Comment (bartoszmajsak) [inline controllers/dscinitialization/servicemesh_setup.go:20]: I am not sure if it wouldn't be better to guard this call the same as we do in `removeServiceMesh` below

```golang
if instance.Spec.ServiceMesh == nil {
	return nil
}
```

The current flow when `instance.Spec.ServiceMesh = nil` is following:

- set `serviceMeshManagementState := operatorv1.Removed`
- call `r.removeServiceMesh(instance)`
- in there we check `if instance.Spec.ServiceMesh == nil` and so we do nothing

Therefore I think we can remove the whole defaulting here and it will have no effect anyway. WDYT?

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
One last thing I missed in the first round. Other than that LGTM

File: controllers/dscinitialization/servicemesh_setup.go (hunk 15 above)
Comment (bartoszmajsak) [reply controllers/dscinitialization/servicemesh_setup.go:20]: Without a pointer we actually face the same issue with "Removed" being now set in DSCI spec. It seems that the intent of "removeServiceMesh" is to only be called from the Finalizer block of DSCI reconcile. This switch case for `operatorv1.Removed` in the original code will also have no effect due to https://github.com/opendatahub-io/opendatahub-operator/blob/ea81e861c457af27c01fa300b3238a44133b0d24/controllers/dscinitialization/servicemesh_setup.go#L55

**Review by bartoszmajsak — APPROVED** [review]
The issue with Removed/Managed execution path is another issue. LGTM

File: controllers/dscinitialization/servicemesh_setup.go (hunk 15 above)
Comment (zdtsw) [reply controllers/dscinitialization/servicemesh_setup.go:20]: we should work on https://issues.redhat.com/browse/RHOAIENG-6984 even it says to "document" but the main purpose is is to define all different cases and correct the logic in the current code base.  e.g should removeServiceMesh be called in any other scenario if it is not deletion of DSCI
